tiledb/sm/array_schema/array_schema_evolution.h

```diff
@@ -0,0 +1,126 @@
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This file defines class ArraySchemaEvolution.
+ */
+
+#ifndef TILEDB_ARRAY_SCHEMA_EVOLUTION_H
+#define TILEDB_ARRAY_SCHEMA_EVOLUTION_H
+
+#include <unordered_map>
+#include <unordered_set>
+
+#include "tiledb/common/status.h"
+#include "tiledb/sm/filter/filter_pipeline.h"
+#include "tiledb/sm/misc/constants.h"
+#include "tiledb/sm/misc/hilbert.h"
+#include "tiledb/sm/misc/uri.h"
+#include "tiledb/sm/misc/uuid.h"
+
+using namespace tiledb::common;
+
+namespace tiledb {
+namespace sm {
+
+class Attribute;
+class Buffer;
+class ConstBuffer;
+class Dimension;
+class Domain;
+class ArraySchema;
+
+enum class ArrayType : uint8_t;
+enum class Compressor : uint8_t;
+enum class Datatype : uint8_t;
+enum class Layout : uint8_t;
+
+/** Specifies the array schema evolution. */
+class ArraySchemaEvolution {
+ public:
+  /* ********************************* */
+  /*     CONSTRUCTORS & DESTRUCTORS    */
+  /* ********************************* */
+
+  /** Constructor. */
+  ArraySchemaEvolution();
+
+  /** Destructor. */
+  ~ArraySchemaEvolution();
+
+  /* ********************************* */
+  /*               API                 */
+  /* ********************************* */
+
+  Status evolve_schema(
+      const ArraySchema* orig_schema, ArraySchema** new_schema);
+
+  /**
+   * Adds an attribute, copying the input.
+   *
+   * @param attr The attribute to be added
+   * @return Status
+   */
+  Status add_attribute(const Attribute* attr);
+
+  /**
+   * Drops an attribute.
+   *
+   * @param attr The attribute to be dropped
+   * @return Status
+   */
+  Status drop_attribute(const std::string& attribute_name);
+
+ private:
+  /* ********************************* */
+  /*         PRIVATE ATTRIBUTES        */
+  /* ********************************* */
+
+  /** The array attributes to be added. */
+  /** It maps each attribute name to the corresponding attribute object. */
+  std::unordered_map<std::string, Attribute*> attributes_to_add_map_;
```

Comment:
Can we switch this to a `tdb_unique_ptr`? We make a copy of the attribute in `add_attribute` and the `array_schema->add_attribute` also makes a copy from the pointer so having a `tdb_unique_ptr` here should be fine.

Reply:
changed raw pointer to tdb_unique_ptr